src/coreComponents/mesh/generators/VTKUtilities.cpp

```diff
@@ -640,6 +640,115 @@ vtkSmartPointer< vtkDataSet > manageGlobalIds( vtkSmartPointer< vtkDataSet > mes
+  globalIndex const numProcs = MpiWrapper::commSize( comm );
+
+  array1d< globalIndex > elemCounts( numProcs );
+  MpiWrapper::allGather( numElems, elemCounts, comm );
+
+  SortedArray< globalIndex > recipientRanks;
+  array1d< globalIndex > donorRanks;
+  array1d< globalIndex > donorNumElems;
+  recipientRanks.reserve( numProcs );
+  donorRanks.reserve( numProcs );
+  donorNumElems.reserve( numProcs );
+
+  for( globalIndex iRank = 0; iRank < numProcs; ++iRank )
+  {
+    if( elemCounts[iRank] == 0 )
+    {
+      recipientRanks.insert( iRank );
+    }
+    else if( elemCounts[iRank] > 1 ) // need at least two elems to be a donor
+    {
+      donorRanks.emplace_back( iRank );
+      donorNumElems.emplace_back( elemCounts[iRank] );
+    }
+  }
+
+  // step 2: at this point, we need to determine the donors and which cells they donate
+
+  // First we sort the donor in order of the number of elems they contain
+  std::stable_sort( donorRanks.begin(), donorRanks.end(),
+                    [&donorNumElems] ( auto & i1, auto & i2 )
+  { return donorNumElems[i1] < donorNumElems[i2]; } );
+
+  // Then, if my position is "i" in the donorRanks array, I will send half of my elems to the i-th recipient
+  globalIndex const myRank = MpiWrapper::commRank();
+  auto const myPosition =
+    LvArray::sortedArrayManipulation::find( donorRanks.begin(), donorRanks.size(), myRank );
+  bool const isDonor = myPosition != donorRanks.size();
+
+  // step 3: my rank was selected to donate cells, let's proceed
+  // we need to make a distinction between two configurations
+
+  array1d< globalIndex > newParts( numElems );
+  newParts.setValues< parallelHostPolicy >( myRank );
+
+  // step 3.1: donorRanks.size() >= recipientRanks.size()
+  // we use a strategy that preserves load balancing
+  if( isDonor && donorRanks.size() >= recipientRanks.size() )
+  {
+    if( myPosition < recipientRanks.size() )
+    {
+      globalIndex const recipientRank = recipientRanks[myPosition];
+      for( globalIndex iElem = std::floor( numElems/2.0 ); iElem < numElems; ++iElem )
+      {
+        newParts[iElem] = recipientRank; // I donate half of my cells
+      }
+    }
+  }
+  // step 3.2: donorRanks.size() < recipientRanks.size()
+  // this is the unhappy path: we don't care anymore about load balancing at this stage
+  // we just want to simulation to run and count on ParMetis/PTScotch to restore load balancing
+  else if( isDonor && donorRanks.size() < recipientRanks.size() )
+  {
+    globalIndex firstRecipientPosition = 0;
+    for( globalIndex iRank = 0; iRank < myPosition; ++iRank )
+    {
+      firstRecipientPosition += elemCounts[iRank] - 1;
+    }
+    if( firstRecipientPosition < recipientRanks.size() )
+    {
+      bool const isLastDonor = myPosition == donorRanks.size() - 1;
+      globalIndex const lastRecipientPosition = firstRecipientPosition + numElems - 2;
+      GEOS_THROW_IF( isLastDonor && ( lastRecipientPosition < recipientRanks.size() - 1 ),
```

Comment:
Sorry for the super minor nit, this might be just a tad easier to read if you treat `lastRecipientPosition` as a typical C++/STL-style one-past-the-end iterator
```suggestion
      globalIndex const lastRecipientPosition = firstRecipientPosition + numElems - 1;
      GEOS_THROW_IF( isLastDonor && ( lastRecipientPosition < recipientRanks.size() ),
```

Reply:
Done

Reply:
Done